src/traces/violin/attributes.js

```diff
@@ -0,0 +1,113 @@
+/**
+* Copyright 2012-2017, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+'use strict';
+
+var boxAttrs = require('../box/attributes');
+var scatterAttrs = require('../scatter/attributes');
+var extendFlat = require('../../lib/extend').extendFlat;
+
+module.exports = {
+    y: boxAttrs.y,
+    x: boxAttrs.x,
+    x0: boxAttrs.x0,
+    y0: boxAttrs.y0,
+    name: boxAttrs.name,
+    orientation: extendFlat({}, boxAttrs.orientation, {
+        description: [
+            'Sets the orientation of the violin(s).',
+            'If *v* (*h*), the distribution is visualized along',
+            'the vertical (horizontal).'
+        ].join(' ')
+    }),
+
+    bandwidth: {
+        valType: 'number',
+        min: 0,
+        role: 'info',
+        editType: 'plot',
+        description: [
+            'Sets the bandwidth used to compute the kernel density estimate.',
+            'By default, the bandwidth is determined by Silverman\'s rule of thumb.'
+        ].join(' ')
+    },
+    scaleby: {
+        valType: 'enumerated',
+        values: ['width', 'area', 'count'],
+        dflt: 'width',
+        role: 'info',
+        editType: 'calc',
+        description: [
+            'Sets the method by which the width of each violin is determined.',
+            '*width* means each violin has the same (max) width',
+            '*area* means each violin has the same area',
+            '*count* means the violins are scaled by the number of sample points making',
+            'up each violin.'
+        ].join('')
+    },
+    span: {
+        valType: 'info_array',
+        items: [
+            {valType: 'any', editType: 'plot'},
+            {valType: 'any', editType: 'plot'}
+        ],
+        role: 'info',
+        editType: 'plot',
+        description: [
+            'Sets the span in data space for which the density function will be computed.',
+            'By default, the span goes from the minimum value to maximum value in the sample.'
+        ].join(' ')
```

Comment:
To address, @alexcjohnson 's concern (from a private convo):

> mmm, looking through example images I don’t actually see the behavior I was imagining… might be something to play with at some point but not for now, unless some standard existing package does it.

> One thing I see a lot in examples that seems to me a *bad* idea is truncating the violin at value where it has a finite width - particularly if that value is actually a data value. Unfortunately one of the examples I see that does exactly that is on our own site… https://plot.ly/python/basic-statistics/ out[5]

> unless there’s some physical limit to the variable… that seems like it would justify doing this, but aren’t you still throwing away area (and therefore visual weight) from the points at the end?

> at any rate, seems like if one is going to truncate like that it should be done explicitly, and should not be the default.

> Ideally I’d like you to only be able to truncate by value (per the argument about physical limits) rather than “truncate at the ends of the data range” but I imagine people will complain if they can’t do both…

Reply:
To note, `span` was inspired [seaborn](http://seaborn.pydata.org/generated/seaborn.violinplot.html)'s `cut` argument.

Reply:
I love that this is two numbers (or date strings, I guess), one for each end. - that's way more flexible than seaborn's one number or ggplot2's boolean. I also love that it's data values, rather than a delta from the ends of the distribution. Between the two of those you could do what seems to be really the *right* thing with seaborn's [cut example](http://seaborn.pydata.org/_images/seaborn-violinplot-12.png) and lop off unphysical values < 0 but not restrict the upper end.

Ideally I would prefer if the default were no clipping at all (which, as discussed, probably means data bounds extended by 2 or 3 times the bandwidth), rather than clipped at the data bounds. The challenge then is to make it easy for people who do want to clip tightly to the data bounds - particularly if we think about the distribution changing over time, having to manually update the bounds for either of these cases seems awkward.

I suppose we could define two special values of `span[i]` - one for "clip tightly to this end of the data" and another for "do not clip this end"? `'tight'` and `'loose'` perhaps? Alternatively there could be another attribute for this, which would be nice as we wouldn't need to mix numbers and strings (or whatever the special value is) but it seems tricky to cover all cases this way, like if you want the low end tight and the upper end loose.

Reply:
> I suppose we could define two special values of span[i]

That would be dangerous, as we want the `span` items to be set in `d` coordinates. So a violin trace with `'tight'` and `'loose'` categories would totally break this.

@alexcjohnson 's `'tight'` and `'loose'` suggestions made me think of https://github.com/plotly/plotly.js/issues/1876 where the specs `axis.bounds` and `axis.boundsmode` are written down.

For consistency, I'm thinking about adding a `spanmode` attribute alongside `span` with possible values `'tight'`, `'soft'`, `'manual'`

```js
{
   spanmode: 'soft',
   span: [null, 10]
   // where the null means pick the 'soft' default value
   // i.e. data min minus 2 bandwidths
}

{
   spanmode: 'hard',
   span: [0, null]
   // where the null means pick the 'hard' default value
   // i.e. the data max.
}
```

We could also add `'tight-soft'` and `'soft-tight'` spanmode values to allow users to pick hard and soft defaults for each ends without having to set `span` at all.

Reply:
Done in https://github.com/plotly/plotly.js/pull/2116/commits/ad51966b2c72b59633415f711349b6836a9cf3e8